README.md
FlindersDSpace-importERA
========================

## Description
ERA is described at http://www.arc.gov.au/era

This suite of applications allows one to:
- obtain appropriate ERA metadata-only files from your univerity's
  Research Management Information System (MIS)
- create a DSpace community/collection structure for the above metadata
- import the Research MIS metadata into the above DSpace structure (without
  duplicating items imported during previous ERA reporting-years)
- map a new item to more than one new collection and map an existing item
  to more than one new and/or existing collection

## Concepts
- An ERA reporting-year is a year in which a "full round of ERA occurred"
  as described on the above web page. So far, these have been ERA 2010
  and ERA 2012. Preparations are currently under way for ERA 2015.

- The target ERA reporting-year is the reporting-year for which ERA
  metadata are being imported into DSpace. Eg. ERA 2012.

- Previous ERA reporting-years are years which have already been imported
  into DSpace prior to the target ERA reporting-year. Eg. If the target
  ERA reporting-year is ERA 2012, then the only previous ERA reporting-year
  would be ERA 2010.

- A DSpace ERA item corresponds to a metadata record for a research output.
  It has an associated identifier, RMID (ie. Research-MIS ID).

- A DSpace ERA collection corresponds to a 4-digit FOR (Field of Research) code.

- The ERA DSpace structure shall comprise the following hierarchical levels.
  * Optional higher level communities. In our case, one top level community
    called "Research Publications".
  * Below that, a community for each ERA reporting-year. Eg. "ERA 2010" and
    "ERA 2012".
  * Below each ERA reporting-year community, a set of sub-communities. In
    our case, we expect these to be either a representation of ERA clusters
    (eg. as specified by the ERA 2012 Discipline Matrix
    http://www.arc.gov.au/era/era_2012/archive/key_documents_2012.htm) or
    a representation of 2-digit FOR codes.
  * Below each cluster or 2-digit FOR sub-community, a set of 4-digit FOR
    collections. Note that 4-digit FOR collection names are likely to be
    repeated from one ERA reporting-year to the next.
  * Below each collection, one item per research output. The items are
    typically metadata-only but may have bitstreams added manually later.

- An item (research output) may belong to one or more collections (FOR codes)
  for the target ERA reporting-year. These collections shall be specified in
  a CSV file.

- An item (identified by its RMID) for the target ERA reporting-year may
  also have appeared in a previous ERA reporting-year (or elsewhere in a
  DSpace collection which is unrelated to ERA). In order to avoid
  duplication of the item and retain any existing bitstreams or metadata
  edits, the newly sent item (which is part of the target ERA reporting-year)
  shall be *discarded* and the existing DSpace item shall be used.

- An item (research output) may belong to one or more collections (FOR codes)
  for previous ERA reporting-years (or elsewhere in a DSpace collection which
  is unrelated to ERA). These collections shall be extracted from the DSpace
  database and placed into a CSV file.

- Any items in the target ERA reporting-year which exist in more than one
  collection (from target or previous ERA reporting-years or elsewhere in
  another DSpace collection) shall be mapped so that they appear in all the
  existing DSpace collections plus the (new) target ERA reporting-year
  collections.


## Warning

Most of the programs in this repository create DSpace Simple Archive Format
(SAF) or Batch Metadata Editing Tool (BMET) files and directories so are
fairly safe to run on your system. However, please exercise great
caution when:
- running the DSpace structure-builder tool
- importing items into DSpace using SAF (eg. the script created by
  bin/erasaf_mkimport.rb)
- updating items in DSpace using BMET (eg. during phases 2 and 3)
- running programs which can potentially make changes to huge
  numbers of (intended or unintended) files eg. utils/del_lang_field.sh

To use this software, it is assumed you have a good understanding
of the DSpace application, Linux/Unix shell and perhaps a little
understanding of Ruby.

Do not use this software unless you know what you are doing.

**Use this software at your own risk.**

## Workflow

More details regarding many of the applications listed below are given
in README*.md files within this directory.

### :bell:  *Phase 1 - Ingest new items*

### Obtain information from your Research MIS
Obtain the following ERA target reporting-year (eg. ERA 2012) information
from your Research MIS
- RM-format (CSV1) CSV file which specifies an RMID per item plus a
  mapping to all corresponding collections for the ERA target reporting-year.
- Item metadata for research outputs in DSpace Simple Archive Format (SAF).
  * Each 4-digit FOR code will have its own collection. SAF
    collection-level directories must be assigned the same name as the
    collection's 4-digit FOR code.
  * SAF item-level directories must be assigned the same name as the item's
    RMID.
  * SAF item-level dublin_core.xml files must contain an RMID-identifier
    as shown below (and DSpace must be configured to recognise such a
    Dublin Core RMID-identifier via the Metadata Registry).
```
<dcvalue element="identifier" qualifier="rmid" ... >...</dcvalue>
```

Note 1: Even if an RMID (ie. item) must appear against several 4-digit
FOR codes (ie. collections) it must only be defined once within the whole
SAF tree (supplied by the Research MIS system). However, it will ultimately
appear in all corresponding collections due to the RM-format (CSV1) CSV
mapping file discussed above (also supplied by the Research MIS system).

Note 2: This process assumes that the Research MIS system does not
know (or chooses to ignore) which RMIDs already exist within DSpace
and hence supplies _all_ RMIDs for the target ERA reporting-year.
However because this suite of programs is DSpace-aware, aspects
regarding duplication and mapping are dealt with by this workflow.

### Decide where to search in DSpace for existing RMIDs

Before or very early in this workflow, you need to decide if you will
look for items with RMIDs (ie. with a populated dc.identifier.rmid field)
from the the whole of DSpace or from only within the top ERA community
(called "Research Publications" in our case) whose children are
ERA reporting-year communities (ie. ERA 2010 and ERA 2012).

This decision affects:
- the app which extracts items with RMIDs from the database ie. use
  prep/itemHdl_colHdl_ResearchPubEra.sh (rather than
  prep/itemHdl_colHdl_AllPub.sh)
- the third command line argument to app bin/rmcsv2hdlcsv.rb (ie.
  use the 2 handles representing the ERA 2010 and ERA 2012
  communities rather than the argument 'any') 

In our case, we decided to only look for items with RMIDs from
under the Research Publications ERA-year communities rather
than any item with an RMID from the whole of DSpace. This is because
our phase 3 steps would be to update some ERA 2010 metadata
fields (ie. dc.type and dc.subject.forgroup) and we would 
prefer to only do this for items which are owned by (or mapped
to) the ERA 2010 community rather than to items owned by others.

(In the workflow test within the test folder, we allowed
all of DSpace to be searched for items with RMIDs.)

### bin/csv2xmlstruct.rb
Create a community/collection DSpace structure to hold items for the ERA
target reporting-year (eg. ERA 2012). The structure is written to an XML
file.

Use the DSpace structure-builder tool to convert the XML file into a
community/collection hierarchy as described in the
[Concepts](#concepts) section above. More information is provided
[here](README20_csv2xmlstruct.md).

### prep/itemHdl_colHdl_ResearchPubEra.sh
Extract all RMIDs in all DSpace ERA reporting year trees in
RMID-Only-format, CSV2. (Alternatively, use
prep/itemHdl_colHdl_AllPub.sh to extract all RMIDs in all of
DSpace in RMID-Only-format, CSV2.) You should extract this
data at this point in the workflow before new items are added
for the target year.

If any RMIDs are null, such item records will be grouped
together at the bottom of this file (due to the SQL 'order
by' clause). This workflow has only been used with such
null-RMID item records removed from this CSV file.

### bin/erasaf_check.rb
Perform some checking on the ERA SAF-tree.

### bin/erasaf_pluckitem.rb
If items are present within the ERA target reporting-year SAF directory tree
but already exist within DSpace from previous ERA reporting-years (or
perhaps exist independently of ERA) then this application "plucks" such
items from the SAF tree so that they will not be imported into DSpace
(as discussed in the [Concepts](#concepts) section above).

In order to prepare for phase 3 (Update old items with new field values)
you should create a CSV file containing only items which have been
plucked out of the SAF-tree. After running bin/erasaf_pluckitem.rb (which
creates the file plucked_items_regex.txt) you can do that as below.
```
# Copy the header line from the earlier database extract
head -1 itemHdl_colHdl_ResearchPubEra.csv > itemHdl_colHdl_ResearchPubEra_pluckedout.csv

# Extract the plucked-out lines from the earlier database extract
egrep -f plucked_items_regex.txt itemHdl_colHdl_ResearchPubEra.csv >> itemHdl_colHdl_ResearchPubEra_pluckedout.csv
```

### bin/erasaf_mkimport.rb
DSpace provides a tool to import a single SAF collection into DSpace.
For the target ERA reporting-year, we wish to import over 150
(4-digit FOR) collections into DSpace. This app creates a shell script
which achieves this by invoking the DSpace import command over 150 times.

Run the resulting shell script to import all (non-duplicate/non-plucked)
target ERA reporting-year items into the newly created ERA structure.

### :bell:  *Phase 2 - Map items to multiple collections*

### bin/rmcsv2hdlcsv.rb
This application can only be run after the target ERA reporting-year
items have been imported into DSpace (and hence have been assigned a
handle).

Convert RM-format (CSV1) CSV file from the Research Management Information
System to a Handle-format (CSV3) CSV file so that collections can be
uniquely identified by their handle (even if they have an identical
collection name such as a 4-digit FOR code under some other ERA
reporting-year).

### bin/erabme_multicollections.rb
For any ERA item for the target ERA reporting year (extracted from the
Research MIS) which we have already imported into DSpace (either in the
target or previous reporting years) this application will create a CSV
file which will map the item to all collections to which it belongs.

As input to this application, you should reuse the CSV file created
above (by running prep/itemHdl_colHdl_AllPub.sh or
prep/itemHdl_colHdl_ResearchPubEra.sh). However this time the
application will use the columns specified for Handle-format, CSV3.

The resulting file (in Handle-format, CSV3) is almost identical to a
Batch Metadata Editing Tool (BMET) CSV file except that items are
identified by their handle rather their item ID.

### bin/hdl2item_bmecsv.rb
This application takes the above Handle-format (CSV3) CSV file
and converts the item-handle into an item-id as required for
the BMET CSV format.

Then we can make other minor changes to the CSV file, in particular:
- rename "col_hdls" column to "collection" column for BMET
- remove unwanted columns (leaving only id and collection columns)
```
# Fix CSV columns for BMET
awk -F\" '$4=="col_hdls"{$4="collection"} {printf "%s,%s\n",$2,$4}' hdl2item_bmecsv.csv > bmet.csv

# Apply item mapping into multiple collections via the DSpace BMET
$HOME/dspace/bin/dspace metadata-import -f bmet.csv |tee metadata-import.log
```


### :bell:  *Phase 3 - Update old items with new field values*

### bin/erasaf_updatedbfield.rb

This program processes items which have already been plucked out of the RMIS
SAF-tree because they already exist in the database. (Hence these
items were not included in the SAF import.) The processing involves
updating certain database fields based on newer information in the
RMIS SAF-tree (if applicable).

The functionality of the program is controlled by a command line
switch which can have the value 'add_forgroups' or 'replace_type'.

- 'add_forgroups' switch: This program will compare each item's
  list of dc.subject.forgroup fields in the DSpace database with
  those in the Research MIS data (in the Simple Archive Format
  (SAF) tree specified by PLUCKED_OUT_DIR). If the RMIS data
  for a particular item contains any additional dc.subject.forgroup
  fields, those extra fields will be added to the database (by
  running this script and importing the resulting CSV file using
  the DSpace Batch Metadata Editing Tool).

- 'replace_type' switch: This program will compare each item's
  dc.type field in the DSpace database with that in the Research
  MIS data (in the Simple Archive Format (SAF) tree specified by
  PLUCKED_OUT_DIR). If the two fields are different, the database
  copy of the field will be replaced with the SAF copy (by
  running this script and importing the resulting CSV file using
  the DSpace Batch Metadata Editing Tool). It expects one
  dc.type field per item.


## CSV types used in the above workflow
### CSV1 (RM) format
CSV1 format is only suitable if the target ERA reporting year is
known (otherwise the 4-digit FOR collection might apply to a
collection in any ERA reporting-year). Hence it is suitable for
the target ERA reporting-year so can be extracted directly from
your Research MIS.
```
RMID,FOR4D_Owner,FOR4D_Others
1222333444,0101,0102||0103
1222333555,0104,
```

### CSV2 (RMID-Only) format
```
RMID
1222333444
1222333555
```

### CSV3 (handle) format
CSV3 format is only suitable once a tree structure has been created
for the particular ERA reporting year(s) and items have been ingested
into such structure(s). Before this time, handles for both items and
collections have not been created. Hence this is *unsuitable* for
extraction from your Research MIS.
```
item_hdl,col_owner_hdl,col_others_hdl
123456789/90,123456789/1111,123456789/1112||123456789/1113
123456789/91,123456789/1114,
```

### DSpace BMET CSV format
Consult the DSpace manual for the Batch Metadata Editing Tool CSV format.

### Note
CSV files discussed in the [Workflow](#workflow)
section must contain the columns specified above. However,
the software in this suite (but not DSpace tools) does permit
CSV files to also contain other columns which will be ignored by the
corresponding software. This means that in one of the cases above, the
same CSV data extracted by prep/itemHdl_colHdl_AllPub.sh (or
prep/itemHdl_colHdl_ResearchPubEra.sh) can be used in two places. That
is, previous items which have an RMID can be deemed to satisfy both:
- RMID-Only-format (CSV2), and
- Handle-format (CSV3)

Hence the hybrid CSV2/CSV3 format would look something like:
```
rmid,item_hdl,col_owner_hdl,col_others_hdl
1222333444,123456789/90,123456789/1111,123456789/1112||123456789/1113
1222333555,123456789/91,123456789/1114,
```

## Utilities

### split command

To divide BMET CSV files into units of 1000 lines or less (as recommended
in the DSpace manual) I found the linux/unix 'split' command very useful.
Type 'man split' on your system for help. The method used was:
- remove the CSV header line from the large file
- split into multiple smaller files
- add the CSV header line to every resultant smaller file

### bin/comm2newparent

This tool is not required for the above workflow, however we used it to
move some ERA 2010 (child) communities under a new parent community to
conform to a new structure required by this suite of applications.

### utils/del_lang_field.sh

This tool is not required for the above workflow, however we used it to
delete the dc.language[.*] field within all SAF-tree dublin_core.xml files.
Exercise great caution if you use this script!

### utils/map_if_bitstream_wrap.sh

This tool is not required for the above workflow, however we used it to
map any items with a bitstream attached from a draft/initial ERA collection
to a "final" ERA collection (where final means that all items within have a
bitstream attached). The script is intended to be run via cron overnight.
It invokes map_if_bitstream.rb.

### prep/itemHdl_colHdl_forgroup_ResearchPubEra.sh

This tool is not required for the above workflow, however we used it to
backup ERA 2010 dc.subject.forgroup information (for later verification
that ERA 2010 4-digit FOR codes have a one-to-one correspondance to ERA 2010
4-digit FOR code collections). We used this script rather than the
output of the Batch Metadata Editing Tool export of the ERA 2010
community because it appears BMET will only export items owned by
collections within the community (whereas this script also includes
items owned outside the community but mapped into collections within
the community).

### prep/*.sh

A variety of scripts for extracting useful and (perhaps) interesting
information from the DSpace database.

Each script contains a brief description (assigned to the "descr"
environment variable).



RELEASE_NOTES
FlindersDSpace-importERA
========================

Release Notes
-------------

v1.0
----

Compatible with DSpace 3.1.

Initial version.  Has been used in a production environment for loading
ERA 2012 metadata into DSpace.

High level features:
- Phase 1 - Ingest new items
- Phase 2 - Map items to multiple collections
- Phase 3 - Update old items with new field values

v1.1
----

Compatible with DSpace 3.1.

Added tool utils/map_if_bitstream_wrap.sh to map any items with a
bitstream attached from a draft/initial ERA collection to a "final"
ERA collection (where final means that all items within have a
bitstream attached)

v1.2
----

Compatible with DSpace 3.1.

Added tool utils/get_item_info.sh

Added tool utils/map_if_fulltext_wrap.sh. Similar to map_if_bitstream_wrap.sh
but the objective is to map any items (in any collection) containing
non-embargoed full-text into a full-text collection. The intent is that
OAI-PMH harvesters can choose to harvest this collection (instead of the
whole of DSpace) in order to only obtain items with full-text. This is
because metadata-only items can be confusing to users. The following
features are included.
- Updates the last_modified field when we map an item with full-text so
  that incremental OAI-PMH behaves properly for the "full-text collection"
  set (particularly for items where an embargo has been lifted).
- Has a MAX_ITEMS_TO_PROCESS parameter.
- Has a IS_DRY_RUN feature.



utils/map_if_bitstream.rb
#!/usr/bin/ruby
#--
# Copyright (c) 2014-2017, Flinders University, South Australia. All rights reserved.
# Contributors: Library, Corporate Services, Flinders University.
# See the accompanying LICENSE file (or http://opensource.org/licenses/BSD-3-Clause).
#
# PURPOSE
#
# Find all items in the source collection (or all collections) which have
# bitstreams and are not yet mapped to the destination collection, then
# create a BMET CSV file which will map them to the destination collection.
#
# ALGORITHM
# - Find all items in the DSpace source collection (or all collections) which
#   have bitstreams and are not yet mapped to the destination collection
# - Create a DSpace Batch Metadata Editing Tool (BMET) CSV file which will
#   map them to the destination collection
# - Send report via email
#
#++
##############################################################################

# Add dirs to the library path
$: << File.expand_path("../lib", File.dirname(__FILE__))
$: << "#{ENV['HOME']}/.ds/etc"

require 'dspace_utils'
require 'resources4bmet_csv'
require 'object_extra'
require 'rubygems'
require 'pg'
require 'pg_extra'
require 'dbc'

##############################################################################
# A class for representing DSpace database items which are to be mapped
# from the source collection to the destination collection.
#
# This program will produce a CSV file suitable for updating the item
# mapping using the DSpace Batch Metadata Editing Tool (BMET).
##############################################################################
class Items4Mapping
  include DSpacePgUtils

  # true = source items from all collections (ie. ignore SOURCE_COLLECTION_HANDLE);
  # false = source items from the collection specified by SOURCE_COLLECTION_HANDLE
  IS_SOURCE_ALL_COLLECTIONS = false				# Customise
  SOURCE_COLLECTION_HANDLE = '123456789/6225'			# Customise
  DEST_COLLECTION_HANDLE   = '123456789/6226'			# Customise

  HANDLE_URL_LEFT_STRING = 'https://dspace.example.com/xmlui/handle/'	# Customise

  WILL_SHOW_RMID = true						# Customise

  MAX_ITEMS_TO_PROCESS = 100			# DSpace 3 manual recommends 1000 BMET records max
  MAX_ITEMS_WARN_MSG = <<-MSG_WARN1.gsub(/^\t*/, '')
	**WARNING**
	  The number of new full-text items has reached the mapping run-limit of #{MAX_ITEMS_TO_PROCESS}.
	  It is recommended that you check that this high-number of items is expected.
  MSG_WARN1

  ############################################################################
  # Constructor for this object
  ############################################################################
  def initialize
    @items = []
  end

  ############################################################################
  # Extract and return items from the database. These items:
  # - have not been withdrawn;
  # - have been published (in_archive = 't' and owning_collection is not null
  #   and exists in handle table);
  # - are in the source collection (ie. SOURCE_COLLECTION_HANDLE) or all
  #   collections;
  # - are not in the destination collection (ie. DEST_COLLECTION_HANDLE);
  # - have (non-licence) bitstreams attached (ie. bundle is 'ORIGINAL')
  #   which have not been deleted and which have at least one bitstream
  #   which is not under embargo
  #
  # Also, these items might be owned-by or mapped-to a collection which is
  # neither SOURCE_COLLECTION_HANDLE nor DEST_COLLECTION_HANDLE.
  ############################################################################
  def get_items_from_db
    sql_source_clause = if IS_SOURCE_ALL_COLLECTIONS
      <<-SQL_GET_ITEMS_FROM_ALL_COLLECTIONS.gsub(/^\t*/, '')
	    item_id in (select item_id from collection2item)
      SQL_GET_ITEMS_FROM_ALL_COLLECTIONS

    else
      <<-SQL_GET_ITEMS_FROM_SOURCE_COLLECTION.gsub(/^\t*/, '')
	    item_id in
	    (select item_id from collection2item where collection_id in
	      (select resource_id from handle where resource_type_id=#{RESOURCE_TYPE_IDS[:collection]} and handle='#{SOURCE_COLLECTION_HANDLE}')
	    )
      SQL_GET_ITEMS_FROM_SOURCE_COLLECTION
    end

    sql_dc_field_select_clause = if Resources4BmetCsv.are_object_names_in_metadatavalue
      "select text_value from metadatavalue where resource_type_id=#{RESOURCE_TYPE_IDS[:item]} and resource_id=i.item_id and metadata_field_id in"
    else
      "select text_value from metadatavalue where item_id=i.item_id and metadata_field_id ="
    end

    sql_bundle_id_in_clause = if Resources4BmetCsv.are_object_names_in_metadatavalue
      <<-SQL_ITEM_ID_IN1.gsub(/^\t*/, '')
              (select resource_id from metadatavalue where text_value='ORIGINAL' and resource_type_id=#{RESOURCE_TYPE_IDS[:bundle]} and metadata_field_id in
                (select metadata_field_id from metadatafieldregistry where element='title' and qualifier is null)
              )
      SQL_ITEM_ID_IN1

    else
      <<-SQL_ITEM_ID_IN2.gsub(/^\t*/, '')
	      (select bundle_id from bundle where name='ORIGINAL' and bundle_id in
	      )
      SQL_ITEM_ID_IN2

    end

    sql = <<-SQL_GET_ITEMS.gsub(/^\t*/, '')
	select
	  i.item_id,
	  i.last_modified,
	  (select handle from handle where resource_type_id=#{RESOURCE_TYPE_IDS[:item]} and resource_id=i.item_id) item_hdl,
	  (select handle from handle where resource_type_id=#{RESOURCE_TYPE_IDS[:collection]} and resource_id=i.owning_collection) owning_collection_hdl,

	  array_to_string(array(
	    select handle from handle where resource_type_id=#{RESOURCE_TYPE_IDS[:collection]} and resource_id in
	      (select collection_id from collection2item where item_id=i.item_id)
	  ), '||') all_collection_hdls,

	  array_to_string(array(
            #{sql_dc_field_select_clause}
	      (select metadata_field_id from metadatafieldregistry where element='identifier' and qualifier='rmid')
	  ), '||') dc_id_rmids,

	  array_to_string(array(
            #{sql_dc_field_select_clause}
	      (select metadata_field_id from metadatafieldregistry where element='title' and qualifier is null)
	  ), '||') dc_titles
	from
	(
	  select distinct item_id, owning_collection, last_modified
	  from item
	  where
	    withdrawn = 'f' and

	    in_archive = 't' and
            owning_collection is not null and
            exists (select resource_id from handle h where h.resource_type_id=#{RESOURCE_TYPE_IDS[:item]} and h.resource_id=item_id) and

            #{sql_source_clause} and

	    item_id not in
	    (select item_id from collection2item where collection_id in
	      (select resource_id from handle where resource_type_id=#{RESOURCE_TYPE_IDS[:collection]} and handle='#{DEST_COLLECTION_HANDLE}')
	    ) and

	    item_id in
            (select item_id from item2bundle where bundle_id in
              #{sql_bundle_id_in_clause}
            and bundle_id in
              (select bundle_id from bundle2bitstream where bitstream_id in
                (select bitstream_id from bitstream where deleted<>'t' and bitstream_id not in
                  (select resource_id from resourcepolicy where resource_type_id=#{RESOURCE_TYPE_IDS[:bitstream]} and start_date > 'now')
                )
              )
            )
	) i
	order by i.item_id;
    SQL_GET_ITEMS

    @items = []
    PG::Connection.connect2(DB_CONNECT_INFO){|conn|
      conn.exec(sql){|result|
        result.each{|row|
          return if @items.length >= MAX_ITEMS_TO_PROCESS

          @items << {
            # Required for BMET CSV file
            :item_id			=> row['item_id'],
            :owning_collection_hdl	=> row['owning_collection_hdl'],
            :all_collection_hdls	=> row['all_collection_hdls'],

            # Only required for reporting
            :last_modified		=> row['last_modified'],
            :item_hdl			=> row['item_hdl'],
            :dc_id_rmids		=> row['dc_id_rmids'],
            :dc_titles			=> row['dc_titles'],
          } 

        }
      }
    }
  end

  ############################################################################
  # Create a BMET CSV string for the items to be mapped (while preserving
  # the owning collection)
  ############################################################################
  def to_s_bmet_csv
    return '' if @items.empty?

    lines = []		# An array of CSV lines
    lines << "id,collection"
    @items.each_with_index{|item, i|
      # First collection must be owner; let second collection be mapped dest
      collection_handles = [ item[:owning_collection_hdl], DEST_COLLECTION_HANDLE ]

      # Append other mapped collections
      all_collection_handles = item[:all_collection_hdls].split(VALUE_DELIMITER).sort
      extra_handles = all_collection_handles - collection_handles
      collection_handles += extra_handles

      lines << "#{item[:item_id]},#{collection_handles.join(VALUE_DELIMITER)}"
    }
    lines.join(NEWLINE) + NEWLINE
  end

  ############################################################################
  # Create a report for these items
  ############################################################################
  def to_s_report
    res = Resources4BmetCsv.new
    preamble = <<-REPORT_PREAMBLE.gsub(/^\t*/, '')
	#{@items.length >= MAX_ITEMS_TO_PROCESS ? MAX_ITEMS_WARN_MSG : ''}
	Program:                        #{File.basename($0)}
	Source collection handle:       #{IS_SOURCE_ALL_COLLECTIONS ? 'All collections' : self.class.handle_to_url(SOURCE_COLLECTION_HANDLE)}
	Source collection name:         #{IS_SOURCE_ALL_COLLECTIONS ? 'All collections' : res.lookup_collection_name(SOURCE_COLLECTION_HANDLE)}

	Destination collection handle:  #{self.class.handle_to_url(DEST_COLLECTION_HANDLE)}
	Destination collection name:    #{res.lookup_collection_name(DEST_COLLECTION_HANDLE)}
	Number of new items to map to destination:  #{@items.length}

    REPORT_PREAMBLE

    lines = []
    unless @items.empty?
      lines << "ITEMS BEING MAPPED:"
      lines << ""

      hdl_width = [
        "#{self.class.handle_to_url(@items.first[:item_hdl])};".length,
        "ItemHandle;".length
      ].max

      lmod_width = [
        "#{@items.first[:last_modified]};".length,
        "LastModified;".length
      ].max

      if WILL_SHOW_RMID
        fmt_str = "%4s) %-11s %-#{hdl_width}s %-6s %-#{lmod_width}s %s"
        lines << sprintf(fmt_str, "#","ItemRMID;","ItemHandle;","Id;","LastModified;","ItemTitle")
        @items.each_with_index{|item, i|
          lines << sprintf(fmt_str, i+1, item[:dc_id_rmids], self.class.handle_to_url(item[:item_hdl]),
            item[:item_id], item[:last_modified], item[:dc_titles])
        }

      else
        fmt_str = "%4s) %-#{hdl_width}s %-6s %-#{lmod_width}s %s"
        lines << sprintf(fmt_str, "#","ItemHandle;","Id;","LastModified;","ItemTitle")
        @items.each_with_index{|item, i|
          lines << sprintf(fmt_str, i+1, self.class.handle_to_url(item[:item_hdl]),
              item[:item_id], item[:last_modified], item[:dc_titles])
        }
      end
    end

    preamble + lines.join(NEWLINE) + NEWLINE
  end

  ############################################################################
  # Convert a handle string into a URL string
  ############################################################################
  def self.handle_to_url(handle)
    "#{HANDLE_URL_LEFT_STRING}#{handle}"
  end

  ############################################################################
  # The main method for this class
  ############################################################################
  def self.main
    STDERR.puts "\nCreating a DSpace BMET CSV file for mapping"
    STDERR.puts   "-------------------------------------------"

    items = Items4Mapping.new
    items.get_items_from_db
    puts items.to_s_bmet_csv
    STDERR.puts items.to_s_report
  end

end

##############################################################################
# Main
##############################################################################
Items4Mapping.main
exit 0

